cpp/cmake_modules/UseCython.cmake

```
@@ -184,4 +184,24 @@ function(cython_add_module _name pyx_target_name generated_files)
   add_dependencies(${_name} ${pyx_target_name})
 endfunction()
 
+execute_process(COMMAND ${PYTHON_EXECUTABLE} -m cython --version
+                OUTPUT_VARIABLE CYTHON_version_output
+                ERROR_VARIABLE CYTHON_version_error
+                RESULT_VARIABLE CYTHON_version_result
+                OUTPUT_STRIP_TRAILING_WHITESPACE
+                ERROR_STRIP_TRAILING_WHITESPACE)
+if(NOT ${CYTHON_version_result} EQUAL 0)
+  set(_error_msg "Command \"${PYTHON_EXECUTABLE} -m cython --version\" failed with")
+  set(_error_msg "${_error_msg} output:\n${CYTHON_version_error}")
+  message(SEND_ERROR "${_error_msg}")
+else()
+  if("${CYTHON_version_output}" MATCHES "^[Cc]ython version ([^,]+)")
+    set(CYTHON_VERSION "${CMAKE_MATCH_1}")
+  else()
+    if("${CYTHON_version_error}" MATCHES "^[Cc]ython version ([^,]+)")
+      set(CYTHON_VERSION "${CMAKE_MATCH_1}")
+    endif()
+  endif()
+endif()
```

Can we simplify this like the following for Cython 0.29.31 or later?

```suggestion
execute_process(COMMAND ${PYTHON_EXECUTABLE} -m cython --version
                OUTPUT_VARIABLE CYTHON_VERSION_OUTPUT
                OUTPUT_STRIP_TRAILING_WHITESPACE)
if(CYTHON_VERSION_OUTPUT MATCHES "^Cython version (.+)")
  set(CYTHON_VERSION "${CMAKE_MATCH_1}")
else()
  set(CYTHON_VERSION "")
endif()
```

---

Should we error out if we fail parsing the output `python -m cython --version`?

---

Thanks for the suggestion @kou! I'm leaving this conversation unresolved due to @pitrou's comment. It makes sense to me to error out if we fail to parse.

---

> Is `-Xfreethreading_compatible=True"` a required flag for Python 3.13?

It's "semi-required" for the free-threaded build of 3.13. More info in [my comment above](https://github.com/apache/arrow/pull/43606#discussion_r1707332850).

---

I'd rather have the CI fail than not notice that some option is disabled.

---

>  It's "semi-required" for the free-threaded build of 3.13.

OK. Can we require Cython 3.1.0a0 or later for Python 3.13 and we always enable the flag with Python 3.13?

> I'd rather have the CI fail than not notice that some option is disabled.

Can we check it by an unit test?
In general, I want to keep CMake related code as simple as possible for easy to maintain.

---

> Can we require Cython 3.1.0a0 or later for Python 3.13 and we always enable the flag with Python 3.13?

We can do that. However, we'll still need CMake code to get the Python version, so there's still going to be some extra CMake stuff, which will probably be the same in terms of maintenance cost as the current version, right?

---

> Can we require Cython 3.1.0a0 or later for Python 3.13

I would prefer to not to that

---

> > I'd rather have the CI fail than not notice that some option is disabled.
> 
> Can we check it by an unit test? In general, I want to keep CMake related code as simple as possible for easy to maintain.

Replacing the `set(CYTHON_VERSION "")` with an error message does not look like an additional maintenance burden, does it?

---

> However, we'll still need CMake code to get the Python version

I think that we can use `Python_VERSION` defined by CMake's `FindPython.cmake` instead of getting it by ourselves: https://cmake.org/cmake/help/latest/module/FindPython.html#result-variables

> Replacing the `set(CYTHON_VERSION "")` with an error message does not look like an additional maintenance burden, does it?

The following is my concern of it: https://github.com/apache/arrow/pull/43606#discussion_r1709715060

> If Cython changes `cython --version format`, PyArrow build will be failed with the latest Cython until we release a new PyArrow version with the latest Cython support.

(FYI: I don't object strict version detection if there are people who will maintain it. If nobody will not do it, I'll do it.)

---

Can we use `${PYTHON_EXECUTABLE} -c "import cython; print(cython.__version__)"` instead of `${PYTHON_EXECUTABLE} -m cython --version` to avoid parsing `cython --version` output?

---

We may want to add `COMMAND_ERROR_IS_FATAL ANY` when we require CMake 3.19 or later: https://cmake.org/cmake/help/latest/command/execute_process.html